lib/ansible/modules/cloud/google/gce_instance_template.py

```diff
@@ -0,0 +1,565 @@
+                if not isinstance(md, dict):
+                    raise ValueError('metadata must be a dict')
+            except ValueError as e:
+                module.fail_json(msg='bad metadata: %s' % str(e))
+            except SyntaxError as e:
+                module.fail_json(msg='bad metadata syntax')
+
+        if hasattr(libcloud, '__version__') and libcloud.__version__ < '0.15':
+            items = []
+            for k, v in md.items():
+                items.append({"key": k, "value": v})
+            metadata = {'items': items}
+        else:
+            metadata = md
+    gce_args['metadata'] = metadata
+
+    if description is not None:
+        gce_args['description'] = description
+
+    instance = None
+    try:
+        instance = gce.ex_get_instancetemplate(name)
+    except ResourceNotFoundError:
+        try:
+            instance = gce.ex_create_instancetemplate(**gce_args)
+            changed = True
+        except GoogleBaseError as err:
+            module.fail_json(
+                msg='Unexpected error attempting to create instance {}, error: {}'
+                .format(
+                    instance,
+                    err.value
+                )
+            )
+
+    if instance:
+        json_data = get_info(instance)
+    else:
+        module.fail_json(msg="no instance template!")
+
+    return (changed, json_data, name)
+
+
+def delete_instance_template(module, gce):
+    """ Delete instance template.
+    module : AnsibleModule object
+    gce: authenticated GCE libcloud driver
+    Returns:
+        instance template information
+    """
+    name = module.params.get('name')
+    current_state = "absent"
+    changed = False
+
+    # get instance template
+    instance = None
+    try:
+        instance = gce.ex_get_instancetemplate(name)
+        current_state = "present"
+    except GoogleBaseError as err:
+        json_data = dict(msg='instance template not exists')
+
+    if current_state == "present":
+        rc = instance.destroy()
+        if rc:
+            changed = True
+        else:
+            module.fail_json(
+                msg='instance template destroy failed'
+            )
+
+    json_data = {}
+    return (changed, json_data, name)
+
+
+def module_controller(module, gce):
+    ''' Control module state parameter.
+    module : AnsibleModule object
+    gce: authenticated GCE libcloud driver
+    Returns:
+        nothing
+    Exit:
+        AnsibleModule object exit with json data.
+    '''
+    json_output = dict()
+    state = module.params.get("state")
+    if state == "present":
+        (changed, output, name) = create_instance_template(module, gce)
+        json_output['changed'] = changed
+        json_output['msg'] = output
+    elif state == "absent":
+        (changed, output, name) = delete_instance_template(module, gce)
+        json_output['changed'] = changed
+        json_output['msg'] = output
+    else:
+        module.fail_json(
```

Comment:
This `else` is not needed, Ansible will verify that only present/absent are passed as the state.